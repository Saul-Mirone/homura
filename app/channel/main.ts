import { IpcMainInvokeEvent } from 'electron';
import Parser from 'rss-parser';
import { Preset } from '../constants/Preset';
import { CreateSourceResult, DB } from '../model';
import { getFaviconByUrl } from '../utils';
import { listenToMain } from './common';

type CheckResult = {
  name: string;
  link: string;
  sourceUrl: string;
  icon?: string;
  items: Array<{
    guid: string;
    title: string;
    link: string;
    content: string;
    date: Date;
  }>;
};

type Event = IpcMainInvokeEvent;
export type Listener = ReturnType<ChannelMain['listener']>;

export class ChannelMain {
  private parser: Parser;

  private checkResult: CheckResult | null;

  constructor(private readonly db: DB) {
    this.parser = new Parser();
    this.checkResult = null;
  }

  public listener() {
    return {
      getSourceList: (_: Event, count: 'unread' | 'starred') =>
        this.getSourceList(count),
      checkUrl: (_: Event, url: string) => this.handleCheckUrl(url),
      confirm: (_: Event, name: string) => this.confirm(name),
      getSourceById: (_: Event, id: number) => this.db.getSourceById(id),
      getPostById: (_: Event, id: number) => this.db.getPostById(id),
      setPostUnread: (_: Event, id: number, unread: boolean) =>
        this.setPostUnread(id, unread),
      setPostStarred: (_: Event, id: number, starred: boolean) =>
        this.setPostStarred(id, starred),
      countBy: (_: Event, type?: 'unread' | 'starred') => this.db.countBy(type),
      getPostByPreset: (_: Event, preset: Preset) =>
        this.db.getPostByPreset(preset),
      markAllAsReadBySourceId: (_: Event, sourceId: number) =>
        this.db.markAllPostsAsReadBySourceId(sourceId),
      sync: () => this.sync(),
      removeSourceById: (_: Event, sourceId: number) =>
        this.db.removeSourceById(sourceId),
      updateSourceNameById: (_: Event, sourceId: number, name: string) =>
        this.db.updateSourceNameById(sourceId, name),
    };
  }

  public listen(): void {
    listenToMain(this.listener());
  }

  private async sync() {
    const urlList = await this.db.getSourceUrlList();
    await Promise.all(
      urlList.map(async ({ sourceUrl, id }) => {
        const { link = '', items = [] } = await this.checkURL(sourceUrl);
        const faviconUrl = await getFaviconByUrl(link);
        const data = {
          link,
          icon: faviconUrl || null,
          posts: items.map((item) => ({
            title: item.title ?? '',
            link: item.link ?? '',
            guid: item.guid ?? item.id ?? item.isoDate ?? '',
            content: item['content:encoded'] ?? item.content ?? '',
            date: new Date(item.isoDate as string),
          })),
        };
        await this.db.diffWithSource(id, data);
      })
    );
  }

  private getSourceList(count: 'unread' | 'starred') {
    return this.db.getSourceList(count);
  }

  private async handleCheckUrl(url: string): Promise<string> {
    try {
      const { title = '', link = '', items = [] } = await this.checkURL(url);
      const faviconUrl = await getFaviconByUrl(link);
      this.checkResult = {
        name: title,
        link,
        sourceUrl: url,
        icon: faviconUrl,
        items: items.map((item) => ({
          title: item.title ?? '',
          link: item.link ?? '',
          guid: item.guid ?? '',
          content: item['content:encoded'] ?? item.content ?? '',
          date: new Date(item.isoDate as string),
        })),
      };
      return title;
    } catch {
      return '';
    }
  }

  private confirm(name: string): Promise<CreateSourceResult> {
    const tmp: CheckResult | null = {
      ...this.checkResult,
    } as CheckResult | null;
    this.checkResult = null;

    if (!tmp) throw new Error();
    return this.db.createSource({
      name,
      link: tmp.link,
      sourceUrl: tmp.sourceUrl,
      icon: tmp.icon || null,
      posts: tmp.items.map(({ title, content, date, guid }) => ({
        guid,
        title,
        content,
        date,
      })),
    });
  }

  private checkURL(url: string) {
    return this.parser.parseURL(url);
  }

  private setPostUnread(id: number, unread: boolean) {
    return this.db.updatePostById(id, {
      unread,
    });
  }

  private setPostStarred(id: number, starred: boolean) {
    return this.db.updatePostById(id, {
      starred,
    });
  }
}
