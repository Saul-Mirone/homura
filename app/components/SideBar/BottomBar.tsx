import React from 'react';
import { PlusIcon, RefreshIcon } from '../Icon';
import { IconContainer } from '../LogoIcon';

export enum Step {
  EnterUrl,
  EnterName,
}

export type BottomBarProps = {
  step: Step | null;
  onClickPlus: () => void;
  onClickSync: () => void;
  render: (step: Step) => JSX.Element;
  refreshing: boolean;
};

export const BottomBar: React.FC<BottomBarProps> = ({
  step,
  onClickPlus,
  onClickSync,
  render,
  refreshing,
}) => (
  <div className="relative flex justify-between text-gray-300">
    {step !== null && (
      <div className="absolute bottom-0 inset-x-0 z-10">{render(step)}</div>
    )}
    <IconContainer onClick={onClickPlus}>
      <PlusIcon />
    </IconContainer>
    <IconContainer onClick={onClickSync}>
      <RefreshIcon spin={refreshing} />
    </IconContainer>
  </div>
);
