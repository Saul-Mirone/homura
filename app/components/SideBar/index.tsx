import React from 'react';

export type SideBarProps = {
  overview: JSX.Element;
  bottom: JSX.Element;
  onClick: () => void;
};

export const SideBar: React.FC<SideBarProps> = ({
  overview,
  bottom,
  children,
  onClick,
}) => {
  return (
    <div
      role="presentation"
      onClick={onClick}
      className="w-1/6 bg-gray-700 pt-10 flex flex-col h-screen select-none"
    >
      <div className="flex-1 thin-scroll">
        {overview}
        <div className="pt-10" />
        {children}
      </div>
      {bottom}
    </div>
  );
};
