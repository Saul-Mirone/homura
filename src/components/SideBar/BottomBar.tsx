import React from 'react';
import { PlusIcon, RefreshIcon } from '../Icon';
import { IconContainer } from '../LogoIcon';

export enum Step {
  EnterUrl,
  EnterName,
}

export type BottomBarProps = {
  step: Step | undefined;
  onClickPlus: () => void;
  onClickSync: () => void;
  render: (step: Step) => JSX.Element;
  loading: boolean;
};

export const BottomBar: React.FC<BottomBarProps> = ({
  step,
  onClickPlus,
  onClickSync,
  render,
  loading,
}) => (
  <div role="toolbar" className="sidebar-bottom-bar">
    {step !== undefined && (
      <div className="sidebar-bottom-bar__step">{render(step)}</div>
    )}
    <IconContainer onClick={onClickPlus}>
      <PlusIcon />
    </IconContainer>
    <IconContainer onClick={onClickSync}>
      <RefreshIcon spin={loading} />
    </IconContainer>
  </div>
);
