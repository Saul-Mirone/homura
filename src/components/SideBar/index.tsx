import React from 'react';
import './style.pcss';

export type SideBarProps = {
    overview: JSX.Element;
    bottom: JSX.Element;
    onClick: () => void;
};

export const SideBar: React.FC<SideBarProps> = ({ overview, bottom, children, onClick }) => (
    <div
        role="menu"
        tabIndex={0}
        aria-label="Sources"
        onClick={onClick}
        onKeyDown={onClick}
        className="sidebar-container"
    >
        <div className="relative flex-1 thin-scroll">
            {overview}
            <div className="pt-10" />
            {children}
        </div>
        {bottom}
    </div>
);
