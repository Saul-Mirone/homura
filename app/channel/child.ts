import { generateChild } from './common';
import { Listener } from './main';

export const channel = generateChild<Listener>();
