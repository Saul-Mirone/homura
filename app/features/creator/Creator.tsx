import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { BottomBar, Step } from '../../components/SideBar/BottomBar';
import { FeedSearchBar } from '../../components/SideBar/FeedSearchBar';
import { FeedSubscribeBar } from '../../components/SideBar/FeedSubscribeBar';
import { AppDispatch } from '../../store';
import {
  confirmName,
  searchUrl,
  selectCreator,
  setLink,
  setName,
  stepToEnterUrl,
} from './creatorSlice';

export const Creator: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const { step, link, name } = useSelector(selectCreator);

  const renderBottom = React.useCallback(
    (currentStep: Step) =>
      currentStep === Step.EnterUrl ? (
        <FeedSearchBar
          link={link}
          onLinkChange={(x) => dispatch(setLink(x))}
          onSearch={() => dispatch(searchUrl())}
        />
      ) : (
        <FeedSubscribeBar
          link={link}
          name={name}
          onNameChange={(x) => dispatch(setName(x))}
          onConfirm={() => dispatch(confirmName())}
        />
      ),
    [dispatch, link, name]
  );

  return (
    <BottomBar
      step={step}
      onClickPlus={() => dispatch(stepToEnterUrl())}
      render={renderBottom}
    />
  );
};
