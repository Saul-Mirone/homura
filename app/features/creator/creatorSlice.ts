import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { channel } from '../../channel/child';
import { Step } from '../../components/SideBar/BottomBar';
import { Mode } from '../../constants/Mode';
import type { AppThunk, RootState } from '../../store';
import { create } from '../source/sourceSlice';

type State = {
  link: string;
  name: string;
  loading: boolean;
  parseError: boolean;
  step: Step | null;
};

const initialState: State = {
  link: '',
  name: '',
  loading: false,
  parseError: false,
  step: null,
};

const creatorSlice = createSlice({
  name: 'creator',
  initialState,
  reducers: {
    stepToEnterUrl: (state) => {
      state.step = Step.EnterUrl;
    },
    stepToEnterName: (state) => {
      state.step = Step.EnterName;
    },
    reset: (state) => {
      state.step = null;
      state.link = '';
      state.name = '';
      state.loading = false;
      state.parseError = false;
    },
    setParserError: (state, action: PayloadAction<boolean>) => {
      state.parseError = action.payload;
    },
    setLink: (state, action: PayloadAction<string>) => {
      state.link = action.payload;
    },
    setName: (state, action: PayloadAction<string>) => {
      state.name = action.payload;
    },
    setLoading: (state, action: PayloadAction<boolean>) => {
      state.loading = action.payload;
    },
  },
});

export const {
  stepToEnterName,
  stepToEnterUrl,
  reset,
  setLink,
  setName,
  setParserError,
  setLoading,
} = creatorSlice.actions;

export const searchUrl = (): AppThunk => async (dispatch, getState) => {
  dispatch(setLoading(true));
  const {
    creator: { link },
  } = getState();

  const result = await channel.checkUrl(link);
  if (!result) {
    dispatch(setLoading(false));
    dispatch(setParserError(true));
    return;
  }

  dispatch(setName(result));
  dispatch(stepToEnterName());
  dispatch(setLoading(false));
};

export const confirmName = (): AppThunk => async (dispatch, getState) => {
  const {
    creator: { name },
    mode,
  } = getState();

  const { posts, icon, link, id } = await channel.confirm(name);

  const rssSource = {
    id,
    name,
    link,
    icon: icon === null ? undefined : icon,
    count: mode === Mode.Starred ? 0 : posts.length,
  };

  dispatch(create(rssSource));

  dispatch(reset());
};

export const creatorReducer = creatorSlice.reducer;

export const selectCreator = (state: RootState) => {
  return {
    ...state.creator,
    refreshing: state.source.refreshing,
  };
};
