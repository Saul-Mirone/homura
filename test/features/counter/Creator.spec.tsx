/* eslint react/jsx-props-no-spreading: off, import/first: off, import/order: off */
import { configureStore } from '@reduxjs/toolkit';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import 'regenerator-runtime/runtime';
import { channel } from '../../../app/channel/child';
import { Step } from '../../../app/components/SideBar/BottomBar';
import { Creator } from '../../../app/features/creator/Creator';
import * as creatorSlice from '../../../app/features/creator/creatorSlice';

jest.mock('../../../app/channel/child');

Enzyme.configure({ adapter: new Adapter() });

function setup(
  preloadedState: {
    creator: {
      link: string;
      name: string;
      step: Step | null;
    };
  } = { creator: { link: '', name: '', step: null } }
) {
  const store = configureStore({
    reducer: { creator: creatorSlice.creatorReducer },
    preloadedState,
  });

  const getWrapper = () =>
    mount(
      <Provider store={store}>
        <Router>
          <Creator />
        </Router>
      </Provider>
    );
  const component = getWrapper();
  return {
    store,
    component,
    buttons: component.find('IconContainer'),
  };
}

describe('Creator component', () => {
  it('should match exact snapshot', () => {
    const { store } = setup();
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router>
            <Creator />
          </Router>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should plus button call increment', () => {
    const { buttons } = setup();
    const stepToEnterUrlSpy = jest.spyOn(creatorSlice, 'stepToEnterUrl');

    buttons.at(0).simulate('click');
    expect(stepToEnterUrlSpy).toBeCalled();
    stepToEnterUrlSpy.mockRestore();
  });
});

describe('Test creator actions', () => {
  it('should call searchUrl', async () => {
    (channel.checkUrl as any).mockResolvedValue('name');
    const fn: any = creatorSlice.searchUrl();
    expect(fn).toBeInstanceOf(Function);
    const dispatch = jest.fn();
    await fn(dispatch, () => ({ creator: { link: 'fake-link' } }));
    expect(dispatch).toBeCalledTimes(2);
  });
});
