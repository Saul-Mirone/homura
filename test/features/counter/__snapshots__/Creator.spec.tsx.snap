// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creator component should match exact snapshot 1`] = `
<div
  className="relative flex justify-between text-gray-300"
>
  <div
    className="p-2 cursor-pointer "
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="w-5 h-5 m-auto"
      role="button"
      tabIndex={0}
    >
      <svg
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 6v6m0 0v6m0-6h6m-6 0H6"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
        />
      </svg>
    </div>
  </div>
  <div
    className="p-2 cursor-pointer "
    role="button"
    tabIndex={0}
  >
    <div
      className="w-5 h-5 m-auto"
      role="button"
      tabIndex={0}
    >
      <svg
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 4v5h.582m15.356 2A8.001 8.001 0 004.582 9m0 0H9m11 11v-5h-.581m0 0a8.003 8.003 0 01-15.357-2m15.357 2H15"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
        />
      </svg>
    </div>
  </div>
</div>
`;
